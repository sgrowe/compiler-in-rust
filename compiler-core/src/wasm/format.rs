use std::io::{self, Write};

#[derive(Default, Debug, Copy, Clone)]
pub struct WasmFormat {
    indent: u32,
}

impl WasmFormat {
    pub fn increase_indent(self) -> Self {
        WasmFormat {
            indent: self.indent + 2,
        }
    }

    pub fn new_line_with_indent<W: Write>(self, writer: &mut W) -> io::Result<()> {
        if self.indent > 0 {
            writer.write_all(b"\n")?;

            for _ in 0..self.indent {
                writer.write_all(b" ")?;
            }
        }

        Ok(())
    }
}
