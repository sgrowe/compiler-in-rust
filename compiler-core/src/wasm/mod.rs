pub use format::WasmTextFormatOptions;
use std::collections::BTreeSet;
use std::fmt::{self, Write};

mod format;

pub trait Wasm<Writer: fmt::Write> {
    fn write_text(&self, writer: &mut Writer, format: WasmTextFormatOptions) -> fmt::Result;
}

#[derive(Debug, Default)]
pub struct WasmModule<'a> {
    functions: Vec<WasmFunction<'a>>,
    exports: Vec<WasmExport<'a>>,
}

impl<'a> WasmModule<'a> {
    pub fn add_function(&mut self, func: WasmFunction<'a>, exported: bool) {
        let name = func.name;

        self.functions.push(func);

        if exported {
            self.exports.push(WasmExport::Function {
                wasm_name: name,
                exported_name: name,
            })
        }
    }
}

impl<'a, Writer: Write> Wasm<Writer> for WasmModule<'a> {
    fn write_text(&self, writer: &mut Writer, format: WasmTextFormatOptions) -> fmt::Result {
        write!(writer, "(module")?;

        let body_format = format.increase_indent();

        for func in &self.functions {
            func.write_text(writer, body_format)?;
        }

        for export in &self.exports {
            export.write_text(writer, body_format)?;
        }

        write!(writer, ")")
    }
}

#[derive(Debug)]
pub struct WasmFunction<'a> {
    name: &'a str,
    params: Vec<&'a str>,
    local_variables: BTreeSet<&'a str>,
    return_type: Option<WasmType>,
    body: Vec<WasmInstruction<'a>>,
}

impl<'a> WasmFunction<'a> {
    pub fn new(
        name: &'a str,
        params: Vec<&'a str>,
        local_variables: BTreeSet<&'a str>,
        return_type: Option<WasmType>,
        body: Vec<WasmInstruction<'a>>,
    ) -> WasmFunction<'a> {
        WasmFunction {
            name,
            params,
            local_variables,
            return_type,
            body,
        }
    }

    pub fn add_local_variable(&mut self, name: &'a str) {
        self.local_variables.insert(name);
    }
}

impl<'a, Writer: Write> Wasm<Writer> for WasmFunction<'a> {
    fn write_text(&self, w: &mut Writer, format: WasmTextFormatOptions) -> fmt::Result {
        format.new_line_with_indent(w)?;

        write!(w, "(func ${}", self.name)?;

        for param in &self.params {
            write!(w, " (param ${} i32)", param)?;
        }

        if let Some(wasm_type) = self.return_type {
            write!(w, " (result {})", wasm_type.to_wasm_text())?;
        }

        for local in &self.local_variables {
            write!(w, " (local ${} i32)", local)?;
        }

        let body_format = format.increase_indent();

        for instruction in &self.body {
            instruction.write_text(w, body_format)?;
        }

        write!(w, ")")
    }
}

#[derive(Debug, Copy, Clone)]
pub enum WasmInstruction<'a> {
    GetLocal(&'a str),
    SetLocal(&'a str),
    ConstI64(i64),
    ConstI32(i32),
    ConstF32(f32),
    AddI64,
    AddI32,
    MinusI32,
    MultiplyI32,
    SignedDivideI32,
    Call(&'a str),
}

impl<'a, Writer: Write> Wasm<Writer> for WasmInstruction<'a> {
    fn write_text(&self, w: &mut Writer, format: WasmTextFormatOptions) -> fmt::Result {
        use WasmInstruction::*;

        format.new_line_with_indent(w)?;

        match self {
            GetLocal(name) => write!(w, "local.get ${}", name),
            SetLocal(name) => write!(w, "local.set ${}", name),
            ConstI64(value) => write!(w, "i64.const {}", value),
            ConstI32(value) => write!(w, "i32.const {}", value),
            ConstF32(value) => write!(w, "f32.const {}", value),
            AddI64 => write!(w, "i64.add"),
            AddI32 => write!(w, "i32.add"),
            MinusI32 => write!(w, "i32.sub"),
            MultiplyI32 => write!(w, "i32.mul"),
            SignedDivideI32 => write!(w, "i32.div_s"),
            Call(name) => write!(w, "call ${}", name),
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum WasmExport<'a> {
    Function {
        wasm_name: &'a str,
        exported_name: &'a str,
    },
}

#[derive(Debug, Copy, Clone)]
pub enum WasmType {
    I32,
    I64,
}

impl WasmType {
    fn to_wasm_text(self) -> &'static str {
        match self {
            WasmType::I64 => "i64",
            WasmType::I32 => "i32",
        }
    }
}

impl<'a, Writer: Write> Wasm<Writer> for WasmExport<'a> {
    fn write_text(&self, writer: &mut Writer, format: WasmTextFormatOptions) -> fmt::Result {
        use WasmExport::*;

        format.new_line_with_indent(writer)?;

        match self {
            Function {
                wasm_name,
                exported_name,
            } => write!(
                writer,
                "(export \"{}\" (func ${}))",
                exported_name, wasm_name
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    fn assert_wasm_output_matches<Input: Wasm<String>>(wasm: Input, expected: &str) {
        let mut out = String::new();

        wasm.write_text(&mut out, WasmTextFormatOptions::default())
            .unwrap();

        assert_eq!(out, expected);
    }

    fn assert_wasm_snapshot_matches<Input: Wasm<String>>(snapshot: &str, wasm: Input) {
        let mut out = String::new();

        wasm.write_text(&mut out, WasmTextFormatOptions::default())
            .unwrap();

        assert_snapshot!(snapshot, out);
    }

    #[test]
    fn formats_empty_function() {
        assert_wasm_output_matches(
            WasmFunction::new("f", vec![], BTreeSet::new(), None, vec![]),
            "(func $f)",
        );
    }

    #[test]
    fn formats_single_arg_function() {
        let func = WasmFunction::new(
            "my_func",
            vec!["arg_1", "arg_2"],
            BTreeSet::new(),
            Some(WasmType::I64),
            vec![],
        );

        assert_wasm_output_matches(
            func,
            "(func $my_func (param $arg_1 i32) (param $arg_2 i32) (result i64))",
        );
    }

    #[test]
    fn formats_module_with_functions() {
        use WasmInstruction::*;
        use WasmType::*;

        let mut module = WasmModule::default();

        module.add_function(
            WasmFunction::new(
                "get_magic_number",
                vec![],
                BTreeSet::new(),
                Some(I64),
                vec![ConstI64(10), ConstI64(5), AddI64],
            ),
            false,
        );

        module.add_function(
            WasmFunction::new(
                "add",
                vec!["arg_1", "arg_2"],
                BTreeSet::new(),
                Some(I64),
                vec![GetLocal("arg_1"), GetLocal("arg_2"), AddI64],
            ),
            true,
        );

        assert_wasm_snapshot_matches("module with two simple functions", module);
    }

    #[test]
    fn formats_simple_export() {
        assert_wasm_output_matches(
            WasmExport::Function {
                exported_name: "add",
                wasm_name: "add_em",
            },
            "(export \"add\" (func $add_em))",
        );
    }
}
