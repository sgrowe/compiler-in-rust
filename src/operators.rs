#[derive(Debug, Copy, Clone)]
pub enum BinaryOperator {
    Plus,
    Minus,
    Multiply,
}
