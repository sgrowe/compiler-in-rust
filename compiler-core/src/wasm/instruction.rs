use super::format::{Wasm, WasmIndentation};
use std::fmt::{self, Write};

#[derive(Debug, Copy, Clone)]
pub enum WasmInstruction<'a> {
    GetLocal(&'a str),
    SetLocal(&'a str),
    ConstI64(i64),
    ConstI32(i32),
    ConstF32(f32),
    AddI64,
    AddI32,
    MinusI32,
    MultiplyI32,
    SignedDivideI32,
    Call(&'a str),
}

impl<'a, Writer: Write> Wasm<Writer> for WasmInstruction<'a> {
    fn write_text(&self, w: &mut Writer, format: WasmIndentation) -> fmt::Result {
        use WasmInstruction::*;

        format.new_line_with_indent(w)?;

        match self {
            GetLocal(name) => write!(w, "local.get ${}", name),
            SetLocal(name) => write!(w, "local.set ${}", name),
            ConstI64(value) => write!(w, "i64.const {}", value),
            ConstI32(value) => write!(w, "i32.const {}", value),
            ConstF32(value) => write!(w, "f32.const {}", value),
            AddI64 => write!(w, "i64.add"),
            AddI32 => write!(w, "i32.add"),
            MinusI32 => write!(w, "i32.sub"),
            MultiplyI32 => write!(w, "i32.mul"),
            SignedDivideI32 => write!(w, "i32.div_s"),
            Call(name) => write!(w, "call ${}", name),
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum WasmType {
    I32,
    I64,
}

impl WasmType {
    pub fn to_wasm_text(self) -> &'static str {
        match self {
            WasmType::I64 => "i64",
            WasmType::I32 => "i32",
        }
    }
}
