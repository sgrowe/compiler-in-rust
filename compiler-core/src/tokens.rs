use super::keywords::Keyword;
use super::operators::BinaryOperator;

#[derive(Debug, Copy, Clone)]
pub enum Token<'a> {
    Equals,
    Pipe,
    Comma,
    Colon,
    FatRightArrow,
    OpenParen,
    CloseParen,
    IndentIncr,
    IndentDecr,
    Name(&'a str),
    Keyword(Keyword),
    BinOp(BinaryOperator),
    Constant(Constant<'a>),
}

#[derive(Debug, Copy, Clone)]
pub enum Constant<'a> {
    Str(&'a str),
    Float(f64),
    Int(i64),
}
