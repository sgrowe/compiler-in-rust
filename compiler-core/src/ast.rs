use super::operators::*;
use super::tokens::*;

#[derive(Debug, Clone)]
pub struct Ast<'a> {
    pub statements: Vec<TopLevelStatement<'a>>,
}

impl<'a> Ast<'a> {
    pub fn new() -> Ast<'a> {
        Ast {
            statements: Vec::new(),
        }
    }

    pub fn append_statement(&mut self, statement: TopLevelStatement<'a>) {
        self.statements.push(statement)
    }
}

#[derive(Debug, Clone)]
pub enum Declaration<'a> {
    Assignment {
        name: &'a str,
        expr: Expression<'a>,
    },
    FunctionDecl {
        name: &'a str,
        arguments: FunctionArgsList<'a>,
        body: Vec<FuncBodyStatement<'a>>,
    },
}

#[derive(Debug, Clone)]
pub enum TopLevelStatement<'a> {
    Declaration {
        decl: Declaration<'a>,
        exported: bool,
    },
}

#[derive(Debug, Clone)]
pub enum FuncBodyStatement<'a> {
    Declaration(Declaration<'a>),
    BareExpression(Expression<'a>),
}

#[derive(Debug, Clone)]
pub struct FunctionArgsList<'a> {
    pub args: Vec<FunctionArg<'a>>,
}

#[derive(Debug, Clone)]
pub struct FunctionArg<'a> {
    pub name: &'a str,
}

#[derive(Debug, Clone)]
pub enum Expression<'a> {
    Variable(&'a str),
    Constant(Constant<'a>),
    FunctionCall {
        name: &'a str,
        args: Vec<Expression<'a>>,
    },
    BinaryOp {
        operator: BinaryOperator,
        left: Box<Expression<'a>>,
        right: Box<Expression<'a>>,
    },
    Negation(Box<Expression<'a>>),
}
