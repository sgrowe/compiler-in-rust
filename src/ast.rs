use super::tokens::*;

#[derive(Debug, Clone)]
pub struct Ast<'a> {
    pub statements: Vec<TopLevelStatement<'a>>,
}

#[derive(Debug, Clone)]
pub enum TopLevelStatement<'a> {
    Assignment { name: &'a str, expr: Expression<'a> },
    BareExpression(Expression<'a>),
}

#[derive(Debug, Clone)]
pub enum Expression<'a> {
    Variable(&'a str),
    Constant(Constant<'a>),
    BinaryOp {
        operator: BinaryOperator,
        left: Box<Expression<'a>>,
        right: Box<Expression<'a>>,
    },
}
