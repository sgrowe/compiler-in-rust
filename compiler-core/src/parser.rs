use super::ast::*;
use super::binding_power::*;
use super::operators::*;
use super::tokeniser::{self, *};
use super::tokens::Token::*;
use super::tokens::*;
use std::iter::Peekable;

pub type Result<'a, X> = std::result::Result<X, ParseError<'a>>;

pub fn parse<'a>(source: &'a str) -> Result<'a, Ast<'a>> {
    Parser::of(tokenise(source)).parse()
}

#[derive(Debug, Clone)]
struct Parser<'a> {
    tokens: Peekable<Tokeniser<'a>>,
}

impl<'a> Parser<'a> {
    fn of(tokens: Tokeniser<'a>) -> Parser<'a> {
        Parser {
            tokens: tokens.peekable(),
        }
    }

    fn step(&mut self) -> tokeniser::Result<Option<Token<'a>>> {
        self.tokens.next().transpose()
    }

    fn peek_next_token(&mut self) -> tokeniser::Result<Option<Token<'a>>> {
        self.tokens.peek().copied().transpose()
    }

    fn parse(&mut self) -> Result<'a, Ast<'a>> {
        let mut ast = Ast::new();

        while let Some(_) = self.peek_next_token()? {
            ast.append_statement(self.top_level_statement(false)?);
        }

        Ok(ast)
    }

    fn top_level_statement(&mut self, is_export: bool) -> Result<'a, TopLevelStatement<'a>> {
        use super::keywords::Keyword::*;

        if let Some(token) = self.step()? {
            match token {
                Name(name) => Ok(TopLevelStatement::Declaration {
                    decl: self.declaration(name)?,
                    exported: is_export,
                }),
                Keyword(Function) => Ok(TopLevelStatement::Declaration {
                    decl: self.function()?,
                    exported: is_export,
                }),
                Keyword(Export) => {
                    if is_export {
                        Err(ParseError::UnexpectedToken(token, "export specified twice"))
                    } else {
                        self.top_level_statement(true)
                    }
                }
                token => Err(ParseError::UnexpectedToken(token, "top level statement")),
            }
        } else {
            Err(ParseError::UnexpectedEndOfInput)
        }
    }

    fn func_body_statement(&mut self) -> Result<'a, FuncBodyStatement<'a>> {
        use super::keywords::Keyword::*;

        if let Some(token) = self.step()? {
            match (token, self.peek_next_token()?) {
                (Name(name), _) => self.named_statement(name),
                (Keyword(Function), _) => self.function().map(FuncBodyStatement::Declaration),
                (Constant(_), _) => Ok(FuncBodyStatement::BareExpression(
                    self.expression(0, Some(token))?,
                )),
                (OpenParen, _) => Ok(FuncBodyStatement::BareExpression(
                    self.expression(0, Some(token))?,
                )),
                (token, _) => Err(ParseError::UnexpectedToken(token, "function body")),
            }
        } else {
            Err(ParseError::UnexpectedEndOfInput)
        }
    }

    fn declaration(&mut self, name: &'a str) -> Result<'a, Declaration<'a>> {
        match self.step()? {
            Some(Equals) => Ok(Declaration::Assignment {
                name,
                expr: self.expression(0, None)?,
            }),
            Some(t) => Err(ParseError::UnexpectedToken(t, "top level assignment")),
            None => Err(ParseError::UnexpectedEndOfInput),
        }
    }

    fn named_statement(&mut self, name: &'a str) -> Result<'a, FuncBodyStatement<'a>> {
        use self::Declaration::*;
        use FuncBodyStatement::*;

        if let Some(Equals) = self.peek_next_token()? {
            self.step()?;

            Ok(Declaration(Assignment {
                name,
                expr: self.expression(0, None)?,
            }))
        } else {
            Ok(BareExpression(self.expression(0, Some(Token::Name(name)))?))
        }
    }

    fn function_call(&mut self, name: &'a str) -> Result<'a, Expression<'a>> {
        let mut args = vec![];

        while let Some(token) = self.step()? {
            match token {
                CloseParen => {
                    return Ok(Expression::FunctionCall { name, args });
                }
                _ => {
                    args.push(self.expression(0, Some(token))?);

                    if let Some(Comma) = self.peek_next_token()? {
                        self.step()?;
                    }
                }
            }
        }

        Err(ParseError::UnexpectedEndOfInput)
    }

    fn function(&mut self) -> Result<'a, Declaration<'a>> {
        match self.step()? {
            Some(Name(name)) => {
                let arguments = self.function_arguments_list()?;
                let body = self.function_body()?;

                Ok(Declaration::FunctionDecl {
                    name,
                    arguments,
                    body,
                })
            }
            _ => Err(ParseError::FunctionParseError),
        }
    }

    fn function_arguments_list(&mut self) -> Result<'a, FunctionArgsList<'a>> {
        match self.step()? {
            Some(OpenParen) => {
                let mut args = Vec::new();

                while let Some(token) = self.step()? {
                    match token {
                        Name(name) => args.push(self.func_arg(name)?),
                        CloseParen => return Ok(FunctionArgsList { args }),
                        _ => return Err(ParseError::ErrorParsingFunctionArgs),
                    }
                }

                Err(ParseError::ErrorParsingFunctionArgs)
            }
            _ => Err(ParseError::ErrorParsingFunctionArgs),
        }
    }

    fn func_arg(&mut self, name: &'a str) -> Result<'a, FunctionArg<'a>> {
        if let Some(Comma) = self.peek_next_token()? {
            self.step()?;
        }

        Ok(FunctionArg { name })
    }

    fn function_body(&mut self) -> Result<'a, Vec<FuncBodyStatement<'a>>> {
        let mut statements = Vec::new();

        match self.step()? {
            Some(IndentIncr) => {}
            _ => return Err(ParseError::IndentExpectedError),
        }

        while let Some(token) = self.peek_next_token()? {
            match token {
                IndentDecr => {
                    self.step()?;
                    return Ok(statements);
                }
                _ => statements.push(self.func_body_statement()?),
            }
        }

        Err(ParseError::UnexpectedEndOfInput)
    }

    fn expression(
        &mut self,
        right_binding_power: u32,
        first_token: Option<Token<'a>>,
    ) -> Result<'a, Expression<'a>> {
        let mut current_token = match first_token {
            Some(t) => t,
            None => self
                .step()?
                .map_or(Err(ParseError::UnexpectedEndOfInput), Ok)?,
        };

        let mut left = self.null_denotation(current_token)?;

        while right_binding_power < self.next_token_binding_power() {
            current_token = self
                .step()?
                .map_or(Err(ParseError::UnexpectedEndOfInput), Ok)?;

            left = self.left_denotation(current_token, left)?;
        }

        Ok(left)
    }

    fn next_token_binding_power(&mut self) -> u32 {
        match self.peek_next_token() {
            Ok(Some(token)) => token.binding_power(),
            _ => 0,
        }
    }

    fn left_denotation(
        &mut self,
        token: Token<'a>,
        left: Expression<'a>,
    ) -> Result<'a, Expression<'a>> {
        match token {
            BinOp(operator) => {
                let right = self.expression(token.binding_power(), None)?;

                Ok(Expression::BinaryOp {
                    operator,
                    left: Box::new(left),
                    right: Box::new(right),
                })
            }
            _ => Ok(left),
        }
    }

    fn null_denotation(&mut self, token: Token<'a>) -> Result<'a, Expression<'a>> {
        match token {
            Constant(c) => Ok(Expression::Constant(c)),
            Name(name) => {
                if let Some(OpenParen) = self.peek_next_token()? {
                    self.step()?;
                    self.function_call(name)
                } else {
                    Ok(Expression::Variable(name))
                }
            }
            BinOp(BinaryOperator::Minus) => {
                Ok(Expression::Negation(Box::new(self.expression(100, None)?)))
            }
            OpenParen => {
                let expr = self.expression(token.binding_power(), None)?;

                match self.step()? {
                    Some(CloseParen) => Ok(expr),
                    Some(token) => Err(ParseError::UnexpectedToken(
                        token,
                        "parsing expression in brackets",
                    )),
                    None => Err(ParseError::UnexpectedEndOfInput),
                }
            }
            _ => Err(ParseError::UnexpectedToken(
                token,
                "null denotation of expression token",
            )),
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum ParseError<'a> {
    TokeniserError(TokeniserError),
    UnexpectedToken(Token<'a>, &'a str),
    UnexpectedEndOfInput,
    FunctionParseError,
    ErrorParsingFunctionArgs,
    IndentExpectedError,
}

impl<'a> From<TokeniserError> for ParseError<'a> {
    fn from(error: TokeniserError) -> Self {
        ParseError::TokeniserError(error)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::fs;
    use test_case::test_case;

    #[test_case("src/fixtures/strings.lang"; "strings")]
    #[test_case("src/fixtures/maths.lang"; "maths")]
    #[test_case("src/fixtures/functions.lang"; "functions")]
    #[test_case("src/fixtures/example_program.lang"; "example program")]
    fn fixtures(fixture_file_name: &str) -> std::io::Result<()> {
        let contents = fs::read_to_string(fixture_file_name)?;

        let ast = parse(&contents);

        assert_debug_snapshot!(ast);

        Ok(())
    }
}
