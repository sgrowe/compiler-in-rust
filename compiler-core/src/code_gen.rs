use super::ast::*;
use super::operators::*;
use super::tokens::*;
use super::wasm::*;
use std::collections::HashSet;

pub fn ast_to_wasm<'a>(ast: &Ast<'a>) -> Result<WasmModule<'a>, CodeGenError> {
    use self::Declaration::*;
    use TopLevelStatement::*;

    let mut module = WasmModule::new();

    for statement in &ast.statements {
        match statement {
            Declaration { decl, exported } => match decl {
                FunctionDecl {
                    name,
                    arguments,
                    body,
                } => {
                    let wasm_args = arguments.args.iter().map(|arg| arg.name).collect();

                    let mut wasm_body = vec![];

                    let mut locals = HashSet::new();

                    for statement in body {
                        let (mut instructions, locals_vars) =
                            compile_func_body_statement(statement)?;

                        locals_vars.iter().for_each(|&name| {
                            locals.insert(name);
                        });

                        wasm_body.append(&mut instructions);
                    }

                    module.add_function(
                        WasmFunction::new(name, wasm_args, locals, Some(WasmType::I32), wasm_body),
                        *exported,
                    )
                }
                Assignment { name: _, expr: _ } => {
                    return Err(CodeGenError::TopLevelAssignmentNotYetSupported)
                }
            },
        }
    }

    Ok(module)
}

#[derive(Debug, Copy, Clone)]
pub enum CodeGenError {
    TopLevelAssignmentNotYetSupported,
    ClosuresNotSupportedYet,
    StringsNotSupportedYet,
}

fn compile_func_body_statement<'a>(
    statement: &FuncBodyStatement<'a>,
) -> Result<(Vec<WasmInstruction<'a>>, Vec<&'a str>), CodeGenError> {
    let mut locals = vec![];

    let instructions = match statement {
        FuncBodyStatement::BareExpression(expr) => compile_expression(expr)?,
        FuncBodyStatement::Declaration(Declaration::Assignment { name, expr }) => {
            let mut instr = compile_expression(expr)?;
            locals.push(*name);
            instr.push(WasmInstruction::SetLocal(name));
            instr
        }
        FuncBodyStatement::Declaration(Declaration::FunctionDecl {
            name: _,
            arguments: _,
            body: _,
        }) => return Err(CodeGenError::ClosuresNotSupportedYet),
    };

    Ok((instructions, locals))
}

fn compile_expression<'a>(expr: &Expression<'a>) -> Result<Vec<WasmInstruction<'a>>, CodeGenError> {
    use self::Constant::*;
    use Expression::*;

    let instructions = match expr {
        Constant(Int(int)) => vec![WasmInstruction::ConstI32(*int as i32)],
        Constant(Float(float)) => vec![WasmInstruction::ConstF32(*float as f32)],
        Constant(Str(_)) => return Err(CodeGenError::StringsNotSupportedYet),
        Variable(name) => vec![WasmInstruction::GetLocal(name)],
        BinaryOp {
            operator,
            left,
            right,
        } => {
            let mut instr = vec![];

            instr.append(&mut compile_expression(left)?);
            instr.append(&mut compile_expression(right)?);
            instr.push(binary_op_to_wasm_instruction(*operator));

            instr
        }
        FunctionCall { name, args } => {
            let mut instr = Vec::with_capacity(args.len() + 1);

            for expr in args {
                instr.append(&mut compile_expression(expr)?);
            }

            instr.push(WasmInstruction::Call(name));

            instr
        }
    };

    Ok(instructions)
}

fn binary_op_to_wasm_instruction<'a>(op: BinaryOperator) -> WasmInstruction<'a> {
    use BinaryOperator::*;
    use WasmInstruction::*;

    match op {
        Plus => AddI32,
        Minus => MinusI32,
        Multiply => MultiplyI32,
    }
}

#[cfg(test)]
mod tests {
    use super::super::parser::parse;
    use super::*;
    use insta::assert_debug_snapshot;
    use std::fs;
    use test_case::test_case;

    #[test_case("src/fixtures/example_program.lang"; "example program")]
    fn fixtures(fixture_file_name: &str) -> std::io::Result<()> {
        let contents = fs::read_to_string(fixture_file_name)?;

        let ast = parse(&contents).unwrap();

        let wasm = ast_to_wasm(&ast);

        assert_debug_snapshot!(wasm);

        Ok(())
    }
}
