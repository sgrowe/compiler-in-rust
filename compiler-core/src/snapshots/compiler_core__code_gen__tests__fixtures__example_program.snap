---
source: compiler-core/src/code_gen.rs
expression: wasm
---
Ok(
    WasmModule {
        functions: [
            WasmFunction {
                name: "add",
                params: [
                    "x",
                    "y",
                ],
                local_variables: {},
                return_type: Some(
                    I32,
                ),
                body: [
                    GetLocal(
                        "x",
                    ),
                    GetLocal(
                        "y",
                    ),
                    AddI32,
                ],
            },
            WasmFunction {
                name: "f",
                params: [
                    "x",
                ],
                local_variables: {
                    "y",
                },
                return_type: Some(
                    I32,
                ),
                body: [
                    GetLocal(
                        "x",
                    ),
                    ConstI32(
                        2,
                    ),
                    AddI32,
                    SetLocal(
                        "y",
                    ),
                    GetLocal(
                        "x",
                    ),
                    GetLocal(
                        "y",
                    ),
                    MultiplyI32,
                ],
            },
            WasmFunction {
                name: "main",
                params: [],
                local_variables: {},
                return_type: Some(
                    I32,
                ),
                body: [
                    ConstI32(
                        2,
                    ),
                    ConstI32(
                        4,
                    ),
                    Call(
                        "add",
                    ),
                    ConstI32(
                        1,
                    ),
                    MinusI32,
                    ConstI32(
                        2,
                    ),
                    MultiplyI32,
                ],
            },
        ],
        exports: [],
    },
)
